src/operator/numpy/linalg/np_norm-inl.h

```diff
@@ -0,0 +1,830 @@
+    } else {  // L0 norm is elementwise constant and non-differentiable
+      ret = 0;
+    }
+    return ret;
+  }
+};
+
+/*! \brief Gradient for abs-min/max */
+struct abs_grad : public mxnet_op::tunable {
+  template<typename DType>
+  MSHADOW_XINLINE static DType Map(DType a, DType b) {
+    DType sgn = DType(sign::Map(a));
+    DType grad = DType(abs::Map(a)) == DType(abs::Map(b)) ?
+                  DType(1.0) : DType(0.0);
+    return sgn * grad;
+  }
+};
+
+/*! \brief Sign */
+struct abs_sign : public mxnet_op::tunable {
+  template<typename DType>
+  MSHADOW_XINLINE static DType Map(DType a, DType b) {
+    return DType(sign::Map(a));
+  }
+};
+
+}  // namespace mshadow_op
+
+inline bool NumpyLpNormShape(const nnvm::NodeAttrs& attrs,
+                             mxnet::ShapeVector *in_attrs,
+                             mxnet::ShapeVector *out_attrs);
+
+inline bool NumpyMatrixNormShape(const nnvm::NodeAttrs& attrs,
+                                   mxnet::ShapeVector *in_attrs,
+                                   mxnet::ShapeVector *out_attrs);
+
+inline void assign_svd_empty(mxnet::ShapeVector *out_attrs);
+
+bool NumpyNormShape(const nnvm::NodeAttrs& attrs,
+                    mxnet::ShapeVector *in_attrs,
+                    mxnet::ShapeVector *out_attrs);
+
+TShape swapMatDims(const TShape &shape, const TShape &axis);
+
+TShape inverseTranspose(const TShape &axes);
+
+struct NumpyNormParam : public dmlc::Parameter<NumpyNormParam> {
+  double ord;
+  dmlc::optional<mxnet::TShape> axis;
+  bool keepdims;
+  int flag;
+  DMLC_DECLARE_PARAMETER(NumpyNormParam) {
+    DMLC_DECLARE_FIELD(ord).set_default(2)
+    .describe("Order of the norm. inf means numpy’s inf object.");
+    DMLC_DECLARE_FIELD(axis).set_default(dmlc::optional<mxnet::TShape>())
+    .describe(R"code(If axis is an integer, it specifies the axis of x along
+     which to compute the vector norms. If axis is a 2-tuple,
+     it specifies the axes that hold 2-D matrices, and the matrix norms of
+     these matrices are computed. If axis is None then either a vector norm (when x is 1-D)
+     or a matrix norm (when x is 2-D) is returned.If axis is an integer,
+     it specifies the axis of x along which to compute the vector norms.
+     If axis is a 2-tuple, it specifies the axes that hold 2-D matrices,
+     and the matrix norms of these matrices are computed. If axis is None then either a
+     vector norm (when x is 1-D) or a matrix norm (when x is 2-D) is returned.)code");
+    DMLC_DECLARE_FIELD(keepdims).set_default(false)
+    .describe("If this is set to `True`, the reduced axis is left "
+    "in the result as dimension with size one.");
+    DMLC_DECLARE_FIELD(flag).set_default(-1)
+    .describe("Mapping relations between ord and flag."
+    "ord:  None,  'fro', 'nuc', 'inf'  '-inf'."
+    "flag:  0 ,    1,      2,    3,      4. ");
+  }
+};
+
+TShape swapMatDims(const TShape &shape, const TShape &axis);
```

Comment:
Seems like you have duplicate declarations of `swapMatDims`, `inverseTranspose` and `NumpyNormShape`.